packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2957,29 +2972,76 @@ export class ContainerRuntime
-				this.mc.logger.sendErrorEvent({
-					eventName: "SignalLost",
-					type: envelope.contents.type,
-					signalsLost: this._perfSignalData.signalsLost,
-					trackingSequenceNumber: this._perfSignalData.trackingSignalSequenceNumber,
-					clientSignalSequenceNumber: envelope.clientSignalSequenceNumber,
-				});
-			} else if (
-				envelope.clientSignalSequenceNumber ===
-				this._perfSignalData.trackingSignalSequenceNumber
+				this._signalTracking.trackingSignalSequenceNumber !== undefined &&
+				this._signalTracking.minimumTrackingSignalSequenceNumber !== undefined
 			) {
-				// only logging for the first connection and the trackingSignalSequenceNUmber.
-				if (this.consecutiveReconnects === 0) {
-					this.sendSignalTelemetryEvent(envelope.clientSignalSequenceNumber);
+				if (
+					envelope.clientSignalSequenceNumber >
+					this._signalTracking.trackingSignalSequenceNumber
+				) {
+					// Incoming signal sequence number is greater than the expected signal sequence number.
+					// So we calculate the number of signals lost and log the event.
+					const signalsLost =
+						envelope.clientSignalSequenceNumber -
+						this._signalTracking.trackingSignalSequenceNumber;
+					this._signalTracking.signalsLost += signalsLost;
+					this.mc.logger.sendErrorEvent({
+						eventName: "SignalLost",
+						type: envelope.contents.type,
+						signalsLost,
+						trackingSequenceNumber: this._signalTracking.trackingSignalSequenceNumber,
+						clientSignalSequenceNumber: envelope.clientSignalSequenceNumber,
+					});
+					// Update the tracking signal sequence number to the next expected signal in the sequence.
+					this._signalTracking.trackingSignalSequenceNumber =
+						envelope.clientSignalSequenceNumber + 1;
+				} else if (
+					envelope.clientSignalSequenceNumber <
+						this._signalTracking.trackingSignalSequenceNumber &&
+					envelope.clientSignalSequenceNumber >=
+						this._signalTracking.minimumTrackingSignalSequenceNumber
+				) {
+					if (
+						this._signalTracking.previousRoundTripSignalSequenceNumber !== undefined &&
+						envelope.clientSignalSequenceNumber >
+							this._signalTracking.previousRoundTripSignalSequenceNumber
+					) {
+						// Log incoming signal that was previously marked as lost as having been received out of order.
+						this._signalTracking.signalsLost--;
+					}
+					this.mc.logger.sendErrorEvent({
+						eventName: "SignalOutOfOrder",
+						type: envelope.contents.type,
+						trackingSequenceNumber: this._signalTracking.trackingSignalSequenceNumber,
+						clientSignalSequenceNumber: envelope.clientSignalSequenceNumber,
+					});
+				} else if (
+					envelope.clientSignalSequenceNumber ===
+					this._signalTracking.trackingSignalSequenceNumber
+				) {
+					// Incoming signal is the next expected signal in the sequence.
+					// Update the tracking signal sequence number to the next expected signal.
+					this._signalTracking.trackingSignalSequenceNumber++;
+				}
```

Comment:
I recommend we switch the order arround here. (1) for performance and readablity we should handle the most common case first. That should be the it is exactly as expected case. (2) the if's can be simplified once the out of order case is last. We'll only need to check >= minTracking as we already know it is less than the expected `trackingSignalSequenceNumber`.
To be really efficient we could write:
```typescript
if (envelope.clientSignalSequenceNumber >= this._signalTracking.trackingSignalSequenceNumber) {
   const signalsLost = envelope.clientSignalSequenceNumber - this._signalTracking.trackingSignalSequenceNumber;
   if (signalsLost > 0) { ... signal lost }
   // Update the tracking signal sequence number to the next expected signal in the sequence.
   this._signalTracking.trackingSignalSequenceNumber = envelope.clientSignalSequenceNumber + 1;
} else if (envelope.clientSignalSequenceNumber >= this._signalTracking.minimumTrackingSignalSequenceNumber) {
   ... out of order
}
```

Reply:
Thanks yeah that looks a lot cleaner. Changed the ordering to reflect this.